theano/contrib/mkl/basic_ops.py

```diff
@@ -737,3 +736,333 @@ def c_code(self, node, name, inp, out, sub):
+        x = T.as_tensor_variable(x)
+        if x.type.ndim != 4:
+            raise TypeError('Input should be a 4-dim variable.')
+        return Apply(self, [x], [MKLNdarrayType(broadcastable=x.type.broadcastable, dtype=x.dtype)()])
+
+    def grad(self, inp, grads):
+        x, = inp
+        gz, = grads
+
+        return [U2IGrad()(x, gz)]
+
+    def c_code(self, node, name, inp, out, sub):
+        x, = inp
+        z, = out
+
+        alpha = self.alpha
+        beta = self.beta
+        k = self.k
+        size = self.size
+
+        if 'float32' == node.inputs[0].type.dtype:
+            precision = 'F32'
+        elif 'float64' == node.inputs[0].type.dtype:
+            precision = 'F64'
+        else:
+            raise TypeError('Type %s is not supported!' % node.inputs[0].type.dtype)
+
+        fail = sub['fail']
+
+        ccode = """
+            int ndim = PyArray_NDIM(%(x)s);
+            int dtype = PyArray_TYPE(%(x)s);
+            npy_intp* d = PyArray_DIMS(%(x)s);
+
+            size_t dims[MNDA_MAX_NDIM] = {0};
+            for (int i = 0; i < ndim; i++) {
+                dims[i] = (size_t)d[i];
+            }
+
+            if (1 == first_run) {
+                bottomSize[0] = d[3]; //w
+                bottomSize[1] = d[2]; //h
+                bottomSize[2] = d[1]; //c
+                bottomSize[3] = d[0]; //n
+
+                bottomStride[0] = 1;
+                bottomStride[1] = bottomStride[0] * bottomSize[0];
+                bottomStride[2] = bottomStride[1] * bottomSize[1];
+                bottomStride[3] = bottomStride[2] * bottomSize[2];
+
+                //create user layout
+                CHECK_ERR( dnnLayoutCreate_%(precision)s(&layout_user, DIMENSION, bottomSize, bottomStride), err );
+                CHECK_ERR( dnnLRNCreateForward_%(precision)s(&primitive, NULL, layout_user, %(size)s, %(alpha)s, %(beta)s, %(k)s), err );
+            }
+
+            if (!( %(z)s
+                && MKLNdarray_Check((PyObject*)%(z)s)
+                && MKLNdarray_NDIM(%(z)s) == ndim
+                && MKLNdarray_DIMS(%(z)s)[0] == d[0]
+                && MKLNdarray_DIMS(%(z)s)[1] == d[1]
+                && MKLNdarray_DIMS(%(z)s)[2] == d[2]
+                && MKLNdarray_DIMS(%(z)s)[3] == d[3]) ) {
+
+                if (%(z)s) Py_XDECREF(%(z)s);
+                %(z)s = (MKLNdarray*)MKLNdarray_New(ndim, dtype);
+                if (!%(z)s) {
+                    %(fail)s;
+                }
+
+                int status = MKLNdarray_set_structure(%(z)s, ndim, dims);
+                if (status != 0) {
+                    %(fail)s;
+                }
+
+                status = MKLNdarray_create_buffer_from_primitive(%(z)s, &primitive, dnnResourceSrc);
+                if (status != 0) {
+                    %(fail)s;
+                }
+            }
+
+            if (!dnnLayoutCompare_%(precision)s(layout_user, MKLNdarray_LAYOUT(%(z)s))) {
+                if (NULL == to_internal) {
```

Comment:
maybe add if (1 == first_run)  like class U2IPool  did in line 610?  
to_internal allocation need to do compare with output %(z)s, but  %(z)s is not in the block of first_run.